### PR TITLE
Enable nullable on vstest.console

### PR DESCRIPTION
**Pull request by Evangelink**
Contributes to [AB#1549371](https://devdiv.visualstudio.com/0bdbc590-a062-4c3f-b0f6-9383f67865ee/_workitems/edit/1549371)

### CODE REVIEW

File: src/vstest.console/CommandLine/CommandLineOptions.cs [hunk 1]
```diff
@@ -156,22 +155,22 @@ public IEnumerable<string> Sources
...
     /// Specifies the Target Device
     /// </summary>
-    public string TargetDevice { get; set; }
+    public string? TargetDevice { get; set; }
 
     /// <summary>
     /// Specifies whether the target device has a Windows Phone context or not
     /// </summary>
-    public bool HasPhoneContext => !string.IsNullOrEmpty(TargetDevice);
+    public bool HasPhoneContext => !StringUtils.IsNullOrEmpty(TargetDevice);
```
Comment (Evangelink) [inline src/vstest.console/CommandLine/CommandLineOptions.cs:173]:
```suggestion
    public bool HasPhoneContext => !TargetDevice.IsNullOrEmpty();
```

File: src/vstest.console/CommandLine/Executor.cs [hunk 2]
```diff
@@ -125,7 +122,7 @@ internal int Execute(params string[] args)
             var diag = Environment.GetEnvironmentVariable("VSTEST_DIAG");
             // This takes Verbose, Info (not Information), Warning, and Error.
             var diagVerbosity = Environment.GetEnvironmentVariable("VSTEST_DIAG_VERBOSITY");
-            if (!string.IsNullOrWhiteSpace(diag))
+            if (!StringUtils.IsNullOrWhiteSpace(diag))
```
Comment (Evangelink) [inline src/vstest.console/CommandLine/Executor.cs:125]:
```suggestion
            if (!diag.IsNullOrWhiteSpace())
```

File: src/vstest.console/CommandLine/InferHelper.cs [hunk 3]
```diff
@@ -47,8 +44,11 @@ public Architecture AutoDetectArchitecture(IList<string> sources, Architecture d
         try
         {
             Architecture? commonArchitecture = null;
-            foreach (string source in sources)
+            foreach (var source in sources)
             {
+                if (source == null)
```
Comment (Evangelink) [inline src/vstest.console/CommandLine/InferHelper.cs:49]: Line 38 we handle null source and here we don't. I have added a skip when null.

File: src/vstest.console/CommandLine/InferHelper.cs [hunk 4]
```diff
@@ -160,15 +161,20 @@ public Framework AutoDetectFramework(IList<string> sources, out IDictionary<stri
...
         sourceToFrameworkMap = new Dictionary<string, Framework>();
 
         var defaultFramework = Framework.DefaultFramework;
-        FrameworkName finalFx = null;
+        FrameworkName? finalFx = null;
         conflictInFxIdentifier = false;
-        foreach (string source in sources)
+        foreach (var source in sources)
         {
+            if (source is null)
```
Comment (Evangelink) [inline src/vstest.console/CommandLine/InferHelper.cs:173]: We are flowing down a list of sources where some items could be null.

File: src/vstest.console/Internal/ConsoleLogger.cs [hunk 5]
```diff
@@ -223,7 +222,7 @@ public void Initialize(TestLoggerEvents events, Dictionary<string, string> param
         }
 
         parameters.TryGetValue(DefaultLoggerParameterNames.TargetFramework, out _targetFramework);
-        _targetFramework = !string.IsNullOrEmpty(_targetFramework) ? NuGetFramework.Parse(_targetFramework).GetShortFolderName() : _targetFramework;
+        _targetFramework = !StringUtils.IsNullOrEmpty(_targetFramework) ? NuGetFramework.Parse(_targetFramework).GetShortFolderName() : _targetFramework;
```
Comment (Evangelink) [inline src/vstest.console/Internal/ConsoleLogger.cs:225]:
```suggestion
        _targetFramework = !_targetFramework.IsNullOrEmpty() ? NuGetFramework.Parse(_targetFramework).GetShortFolderName() : _targetFramework;
```

File: src/vstest.console/Internal/ConsoleLogger.cs [hunk 6]
```diff
@@ -257,22 +256,23 @@ private static void PrintTimeSpan(TimeSpan timeSpan)
     /// </summary>
     private static string GetFormattedOutput(Collection<TestResultMessage> testMessageCollection)
     {
-        if (testMessageCollection != null)
```
Comment (Evangelink) [inline src/vstest.console/Internal/ConsoleLogger.cs:260]: Reverted the condition to reduce nesting.

File: src/vstest.console/Internal/ConsoleLogger.cs [hunk 7]
```diff
@@ -257,22 +256,23 @@ private static void PrintTimeSpan(TimeSpan timeSpan)
...
-                {
-                    sb.AppendFormat(CultureInfo.CurrentCulture, "{0}{1}", TestMessageFormattingPrefix, messageText);
-                }
+        var sb = new StringBuilder();
+        foreach (var message in testMessageCollection)
+        {
+            var prefix = string.Format(CultureInfo.CurrentCulture, "{0}{1}", Environment.NewLine, TestMessageFormattingPrefix);
+            var messageText = message.Text?.Replace(Environment.NewLine, prefix).TrimEnd(TestMessageFormattingPrefix.ToCharArray());
+
+            if (!StringUtils.IsNullOrWhiteSpace(messageText))
```
Comment (Evangelink) [inline src/vstest.console/Internal/ConsoleLogger.cs:270]:
```suggestion
            if (!messageText.IsNullOrWhiteSpace())
```

File: src/vstest.console/Internal/ConsoleLogger.cs [hunk 8]
```diff
@@ -290,20 +290,21 @@ private static Collection<TestResultMessage> GetTestMessages(Collection<TestResu
...
     {
+        TPDebug.Assert(result != null, "a null result can not be displayed");
+        TPDebug.Assert(Output != null, "Initialize should have been called.");
 
         // Add newline if it is not in given output data.
         var addAdditionalNewLine = false;
 
-        Debug.Assert(result != null, "a null result can not be displayed");
-        if (!string.IsNullOrEmpty(result.ErrorMessage))
+        if (!StringUtils.IsNullOrEmpty(result.ErrorMessage))
```
Comment (Evangelink) [inline src/vstest.console/Internal/ConsoleLogger.cs:299]:
```suggestion
        if (!result.ErrorMessage.IsNullOrEmpty())
```

File: src/vstest.console/Internal/ConsoleLogger.cs [hunk 9]
```diff
@@ -290,20 +290,21 @@ private static Collection<TestResultMessage> GetTestMessages(Collection<TestResu
...
         {
             addAdditionalNewLine = true;
             Output.Information(false, ConsoleColor.Red, string.Format("{0}{1}", TestResultPrefix, CommandLineResources.ErrorMessageBanner));
             var errorMessage = string.Format(CultureInfo.CurrentCulture, "{0}{1}{2}", TestResultPrefix, TestMessageFormattingPrefix, result.ErrorMessage);
             Output.Information(false, ConsoleColor.Red, errorMessage);
         }
 
-        if (!string.IsNullOrEmpty(result.ErrorStackTrace))
+        if (!StringUtils.IsNullOrEmpty(result.ErrorStackTrace))
```
Comment (Evangelink) [inline src/vstest.console/Internal/ConsoleLogger.cs:307]:
```suggestion
        if (!result.ErrorStackTrace.IsNullOrEmpty())
```

File: src/vstest.console/Internal/ConsoleLogger.cs [hunk 10]
```diff
@@ -317,7 +318,7 @@ private static void DisplayFullInformation(TestResult result)
             addAdditionalNewLine = true;
             var stdOutMessages = GetFormattedOutput(stdOutMessagesCollection);
 
-            if (!string.IsNullOrEmpty(stdOutMessages))
+            if (!StringUtils.IsNullOrEmpty(stdOutMessages))
```
Comment (Evangelink) [inline src/vstest.console/Internal/ConsoleLogger.cs:321]:
```suggestion
            if (!stdOutMessages.IsNullOrEmpty())
```

File: src/vstest.console/Internal/ConsoleLogger.cs [hunk 11]
```diff
@@ -330,7 +331,7 @@ private static void DisplayFullInformation(TestResult result)
             addAdditionalNewLine = false;
             var stdErrMessages = GetFormattedOutput(stdErrMessagesCollection);
 
-            if (!string.IsNullOrEmpty(stdErrMessages))
+            if (!StringUtils.IsNullOrEmpty(stdErrMessages))
```
Comment (Evangelink) [inline src/vstest.console/Internal/ConsoleLogger.cs:334]:
```suggestion
            if (!stdErrMessages.IsNullOrEmpty())
```

File: src/vstest.console/Internal/ConsoleLogger.cs [hunk 12]
```diff
@@ -343,7 +344,7 @@ private static void DisplayFullInformation(TestResult result)
             addAdditionalNewLine = false;
             var dbgTrcMessages = GetFormattedOutput(dbgTrcMessagesCollection);
 
-            if (!string.IsNullOrEmpty(dbgTrcMessages))
+            if (!StringUtils.IsNullOrEmpty(dbgTrcMessages))
```
Comment (Evangelink) [inline src/vstest.console/Internal/ConsoleLogger.cs:347]:
```suggestion
            if (!dbgTrcMessages.IsNullOrEmpty())
```

File: src/vstest.console/Internal/ConsoleLogger.cs [hunk 13]
```diff
@@ -356,7 +357,7 @@ private static void DisplayFullInformation(TestResult result)
             addAdditionalNewLine = false;
             var addnlInfoMessages = GetFormattedOutput(addnlInfoMessagesCollection);
 
-            if (!string.IsNullOrEmpty(addnlInfoMessages))
+            if (!StringUtils.IsNullOrEmpty(addnlInfoMessages))
```
Comment (Evangelink) [inline src/vstest.console/Internal/ConsoleLogger.cs:360]:
```suggestion
            if (!addnlInfoMessages.IsNullOrEmpty())
```

**Review by nohwnd — APPROVED** [review]

**Pull request merged**